Add new blog post: Demystifying OpenTelemetry: Why You Shouldn’t Fear Observability in Traditional Environments (#8549)
open-telemetry/opentelemetry.io: content/en/blog/2026/demystifying-opentelemetry/index.md

---
title:
  'Demystifying OpenTelemetry: Why You Shouldn’t Fear Observability in
  Traditional Environments'
author: '[Lukasz Ciukaj](https://github.com/luke6Lh43) (Splunk)'
linkTitle: Demystifying OpenTelemetry
date: 2026-01-03
issue: 8548
sig: End-User
cSpell:ignore: ciukaj lukasz
---

For decades, traditional technology environments, ranging from on-premises data
centers to legacy applications and industrial control systems, have powered the
core of many organizations. These systems are battle-tested and deeply woven
into business operations, but they also present unique challenges when it comes
to modernizing IT practices, especially observability.

**Challenges of implementing observability in traditional environments:**

- Noisy, unstructured logs make it hard to extract meaningful information.
- Siloed monitoring data across different tools or systems leads to fragmented
  visibility.
- Limited instrumentation in legacy apps and systems hinders collection of
  modern metrics and traces.
- Concerns about performance impact from adding new observability tooling.
- Integration difficulties when bridging legacy protocols or hardware with
  modern platforms.

<br>

> 

Reviewer: ```suggestion
```

that simply spit out line after line of plain text logs to a file. No JSON, no
structure, no API - just endless lines of text. It’s easy to assume there’s no
way to extract meaningful insight from that mess.

**Why this myth persists?**

In many traditional environments, whether it’s a production line, a legacy
application, or an industrial control system, the only digital “signal” you
might see is a stream of raw, unstructured log files. To an Operations Manager,
these files are frustratingly opaque. She or he cares deeply about specific
faults; knowing whether Line 1 is down due to a "Jam" versus "LowPressure"
dictates immediate response and maintenance strategy. But when that critical
data is buried in unstructured text like **FAULT_DETECTED: Line1, Fault=Jam**,
it is invisible to standard monitoring dashboards. You can't graph text, you
can't alert on a string inside a file easily, and you certainly can't see trends
over time. This leads to the myth that these systems are "black boxes". But with
modern observability tools like OpenTelemetry, these “useless” logs can become a
goldmine of operational insight.

> **Example legacy log lines:**

Reviewer: ```suggestion
### Example legacy log lines
```

```console
2026-01-04 00:39:58 | PRODUCT_COMPLETED: Line1, Count=1
2026-01-04 00:40:00 | FAULT_DETECTED: Line2, Fault=LowPressure
2026-01-04 00:40:02 | MACHINE_START: Line2
2026-01-04 00:40:07 | FAULT_DETECTED: Line2, Fault=Overheat
2026-01-04 00:40:10 | MACHINE_START: Line2
2026-01-04 00:40:14 | PRODUCT_COMPLETED: Line1, Count=1
2026-01-04 00:40:18 | MACHINE_START: Line2
2026-01-04 00:40:21 | PRODUCT_COMPLETED: Line1, Count=1
2026-01-04 00:40:27 | SENSOR_READING: Line1, Temp=83.9
2026-01-04 00:40:29 | FAULT_DETECTED: Line1, Fault=LowPressure
2026-01-04 00:40:32 | SENSOR_READING: Line1, Temp=84.7
2026-01-04 00:40:34 | PRODUCT_COMPLETED: Line1, Count=1
```

**How to make this system observable?**

OpenTelemetry Collector can watch these files in real time, parse the events,
and, without requiring any code changes to the legacy application, transform
them into structured metrics.

> **Sample OpenTelemetry Collector config:**

Reviewer: ```suggestion
#### Sample OpenTelemetry Collector config
```

receivers:
  filelog:
    include: [/logs/legacy.log]
    start_at: end
    operators:
      # 1. Generic Parsing: Capture Timestamp, Event, Line, and put the rest in 'params'
      - type: regex_parser
        regex:
          '^(?P<timestamp>.+?) \| (?P<event_type>[A-Z_]+): (?P<line>Line\d+)(?:,
          (?P<params>.*))?'
        timestamp:
          parse_from: attributes.timestamp
          layout: '%Y-%m-%d %H:%M:%S'

      # 2. Specific Extraction: Look for "Fault=" only inside the 'params'
      - type: regex_parser
        regex: 'Fault=(?P<fault>\w+)'
        parse_from: attributes.params
        if: 'attributes.params != nil'

connectors:
  count:
    logs:
      machine_events_total:
        description: 'Count of manufacturing events by type, line, and fault.'
        attributes:
          - key: event_type
            default_value: 'unknown'
          - key: line
            default_value: 'unknown'
          - key: fault
            default_value: 'none' # Automatically applied if no fault is found

service:
  pipelines:
    logs:
      receivers: [filelog]
      exporters: [count]
    metrics/generated:
      receivers: [count]
      exporters: [prometheus]
```

> **How it works:**

Reviewer: ```suggestion
#### How it works
```

- **Chained Parsing:** The filelog receiver first identifies the high-level
  event (e.g., FAULT_DETECTED). It then runs a second, specific check just to
  extract the Fault type (like "Jam" or "Overheat"). This makes the
  configuration robust and easy to read.
- **Metric Generation:** The count connector converts these parsed logs into a

>   metric called machine_events_total.

Reviewer: ```suggestion
  metric called `machine_events_total`.
```

<br>

> 

Author: Fixed in last commit :)

**Result:**

With this configuration, your old text logs become a structured, queryable data
source. Your Operations Manager can now open a dashboard and see exactly how
many "Jam" faults occurred on "Line 1" in the last hour, fueling data-driven
decisions without changing a single line of code in the legacy app. Myth busted!

Sample dashboard in Prometheus:

> ![Alt text for the image](prometheus-faults-dashboard.png)

Reviewer: Please add alt text for the image or remove the placeholder.

## Myth 2: Our IoT devices publish telemetry to MQTT broker, so integrating with OpenTelemetry isn’t possible.

Our production line relies on robotic arms and sensors that send readings to an
MQTT broker, an industry standard for IoT, but not something OpenTelemetry
natively understands. Does that mean we’re stuck without modern monitoring?

**Why this myth persists:**

MQTT is the messaging backbone for countless industrial and IoT environments,
reliably ferrying sensor data from devices to brokers. However, since MQTT uses
its own lightweight protocol and ecosystem, many teams assume their sensor data
can’t be easily brought into modern observability pipelines. Some MQTT brokers
now natively integrate with OpenTelemetry, allowing direct export of metrics and
traces using the OTLP protocol. If you’re using a modern broker with this
feature, you can simply point your broker at your collector’s OTLP endpoint - no
additional code required. If your broker does not support OTLP export, you’re
still not blocked: you can use a lightweight bridge service to subscribe to MQTT

> topics and forward messages to the OpenTelemetry Collector.

Author: This is good suggestion @danielgblanco ! I updated bottom line paragraph in this myth and included your comment. Thanks for suggesting it!

**Example: Data Sent from an IoT Sensor:**

In our case a payload published by a robotic arm sensor to MQTT might look like:

```json
{
  "device_id": "robot-arm-7",
  "job_id": "abc123",
  "temp": 78.4,
  "humidity": 32.6,
  "job_start": "2025-12-19T12:00:02Z",
  "job_end": "2025-12-19T12:00:05Z"
}
```

This message tells us which device sent it, details about the job, and the
relevant sensor readings.

> **Creating Traces and Spans in the MQTT Bridge App:**

Reviewer: ```suggestion
### Creating traces and spans in the MQTT bridge app
```

> To get real end-to-end visibility (not just metrics) it's powerful to create an

Reviewer: ```suggestion
To get real end-to-end visibility (not just metrics), create an
```

tracer = trace.get_tracer(__name__)
span_processor = BatchSpanProcessor(OTLPSpanExporter(endpoint="http://collector:4318/v1/traces"))
trace.get_tracer_provider().add_span_processor(span_processor)

def on_message(client, userdata, msg):
    payload = json.loads(msg.payload.decode())
    job_start = datetime.datetime.fromisoformat(payload["job_start"].replace("Z", "+00:00"))
    job_end = datetime.datetime.fromisoformat(payload["job_end"].replace("Z", "+00:00"))
    duration = (job_end - job_start).total_seconds()

    with tracer.start_as_current_span(
        "robotic_job",
        start_time=job_start.timestamp(),
        end_on_exit=True,
    ) as span:
        span.set_attribute("device_id", payload["device_id"])
        span.set_attribute("job_id", payload["job_id"])
        span.set_attribute("temperature", payload["temp"])
        span.set_attribute("humidity", payload["humidity"])
        span.set_attribute("job_duration_s", duration)
        # ...additional processing...

# Set up MQTT client
client = mqtt.Client()
client.on_message = on_message
client.connect("mqtt-broker", 1883)
client.subscribe("production/robot-arms")
client.loop_forever()
```

Sample span in Jaeger:

> ![Alt text for the image](sample-span-jaeger.png)

Reviewer: Update alt text or remove the placeholder. Thanks!

Author: Yes, this is fixed in last commit :)

> **What’s the trick here?**

Reviewer: ```suggestion
#### What’s the trick here?
```

By explicitly specifying start_time=job_start.timestamp() (and optionally
end_time), the span precisely tracks the job’s real-world execution, even if the
message is processed later. This gives you accurate, queryable traces that show
exactly when each job occurred and how long it took across devices, processing
steps, and backends.

You have several options for translating IoT sensor data into metrics for
dashboards and alerts:

- **Emit metrics directly from your bridge app:** You can use OpenTelemetry’s
  metrics API to send custom metrics (such as temperature, humidity, or job
  duration) alongside or instead of spans.

> 

Author: Fixed in last commit :)

- **Write a dedicated processor:** Build a custom OpenTelemetry Collector
  processor that derives metrics from incoming spans and then extracting values
  from span attributes.
- **Leverage your observability backend:** Many modern backends can generate
  metrics from span attributes, making it easy to turn your job telemetry into
  actionable, queryable metrics with minimal extra plumbing.

<br>

> 

Reviewer: ```suggestion
```

> **Bottom line:** If your MQTT broker supports OpenTelemetry, use native OTLP

Reviewer: ```suggestion
### Bottom line

If your MQTT broker supports OpenTelemetry, use native OTLP
```

export for seamless integration. If not, a simple bridge app can transform your
sensor and event streams into full observability data. Modern observability
backends make it even easier by allowing metrics to be derived from span
attributes, so you can go from IoT signal to meaningful insight with very little
friction. Myth busted!

> ## Myth 3: Windows and SQL Server Environments Are Incompatible with Observability.

Reviewer: ```suggestion
## Myth 3: Windows and SQL Server environments are incompatible with observability.
```

Windows machines and SQL Servers are the backbone of our operations, running
everything from analytics to inventory. Yet many believe these platforms are
simply out of reach for modern, open observability tooling.

> **Why this myth persists:** It’s a common belief that monitoring and

Reviewer: ```suggestion
### Why this myth persists

It’s a common belief that monitoring and
```

observability are only possible in cloud native or Linux-based systems, leaving
classic Windows servers and SQL Server workloads out of reach. In reality,
OpenTelemetry Collector supports both environments with dedicated receivers that
require minimal configuration. Let’s break it down.

**Observing SQL Server with the OpenTelemetry Collector:**

Many organizations rely on SQL Server databases for production, analytics, or
inventory. With the OpenTelemetry Collector’s sqlserver receiver, you can scrape
health and performance metrics directly, without needing agents on your database
hosts. Below is a sample configuration showing how to set this up:

```yaml
receivers:
  sqlserver/sql1:
    collection_interval: 30s
    username: oteluser
    password: YourStrong!Passw0rd
    server: sql-server-1
    port: 1433

  sqlserver/sql2:
    collection_interval: 30s
    username: oteluser
    password: YourStrong!Passw0rd
    server: sql-server-2
    port: 1433

service:
  pipelines:
    metrics/regular:
      receivers: [sqlserver/sql1, sqlserver/sql2]
      exporters: [prometheus]
```

> **What this achieves:**

Reviewer: ```suggestion
#### What this achieves
```

The Collector regularly scrapes key SQL Server metrics (connections, buffer
pool, locks, batch rates, and more), exposing them to observability backends.

> ![Alt text for the image](prometheus-sqlserver.png)

Reviewer: Alt text, again. 😃

Author: Yes, this is fixed in last commit :)

> **Observing Windows Machines with the Windows Performance Counters Receiver:**

Reviewer: ```suggestion
### Observing Windows machines with the Windows performance counters receiver
```

Classic Windows hosts still drive many production and control environments. The

> Windows Performance Counters Receiver (part of the OpenTelemetry Collector

Reviewer: ```suggestion
[Windows performance counters receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/windowsperfcountersreceiver) (part of the OpenTelemetry Collector
```

Contrib distribution) lets you collect a wide array of system, application, or
custom metrics right from the Windows registry using the native PDH interface.

> Below is a sample configuration for a lightweight agent running on a Windows

Reviewer: ```suggestion
The following is a sample configuration for a lightweight agent running on a Windows
```
